### PR TITLE
Oom fixes 2022.01.19

### PR DESCRIPTION
**Pull request by iulianpascalau**
- made the db pointer inside the storers implementation nil whenever we call Close or Destroy. Added nil checks on all usages of the db pointer.
- peer accounts trie snapshot for shard nodes so the pruning storer for the peer accounts will not cause OOMs
- minor storers refactor: used context instead of chan

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: go.mod [hunk 1]
```diff
@@ -47,6 +47,7 @@ require (
 	golang.org/x/crypto v0.0.0-20210711020723-a769d52b0f97
 	golang.org/x/net v0.0.0-20210428140749-89ef3d95e781
 	gopkg.in/go-playground/validator.v8 v8.18.2
+// test point 3 for custom profiler
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply go.mod:50]: no, will stay

File: storage/leveldb/common.go [hunk 2]
```diff
@@ -69,7 +70,25 @@ func openOneTime(path string, options *opt.Options) (*leveldb.DB, error) {
 }
 
 type baseLevelDb struct {
-	db *leveldb.DB
+	mutDb sync.RWMutex
+	db    *leveldb.DB
+}
+
+func (bldb *baseLevelDb) getDbPointer() *leveldb.DB {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply storage/leveldb/common.go:77]: no code improvement as the checks still remain. Will use slightly more memory (for the error pointer)

**Review by bogdan-rosianu — APPROVED** [review]

File: storage/leveldb/leveldb.go [hunk 3]
```diff
@@ -57,32 +59,34 @@ func NewDB(path string, batchDelaySeconds int, maxBatchSize int, maxOpenFiles in
...
-	go dbStore.batchTimeoutHandle()
+	go dbStore.batchTimeoutHandle(ctx)
 
 	runtime.SetFinalizer(dbStore, func(db *DB) {
 		_ = db.Close()
 	})
 
-	log.Debug("opened level db persister", "path", path)
+	crtCounter := atomic.AddUint32(&loggingDBCounter, 1)
+	log.Debug("opened level db persister", "path", path, "created pointer", fmt.Sprintf("%p", bldb.db), "global db counter", crtCounter)
```
Comment (bogdan-rosianu) [inline storage/leveldb/leveldb.go:84]: could have used `logDebug`

File: storage/leveldb/leveldbSerial.go [hunk 4]
```diff
@@ -83,7 +80,8 @@ func NewSerialDB(path string, batchDelaySeconds int, maxBatchSize int, maxOpenFi
 		_ = db.Close()
 	})
 
-	log.Debug("opened serial level db persister", "path", path)
+	crtCounter := atomic.AddUint32(&loggingDBCounter, 1)
+	log.Debug("opened serial level db persister", "path", path, "created pointer", fmt.Sprintf("%p", bldb.db), "global db counter", crtCounter)
```
Comment (bogdan-rosianu) [inline storage/leveldb/leveldbSerial.go:84]: `logDebug` ?

File: storage/leveldb/leveldb.go [hunk 5]
```diff
@@ -57,32 +59,34 @@ func NewDB(path string, batchDelaySeconds int, maxBatchSize int, maxOpenFiles in
...
 
 	return dbStore, nil
 }
 
-func (s *DB) batchTimeoutHandle() {
+func (s *DB) batchTimeoutHandle(ctx context.Context) {
 	for {
 		select {
 		case <-time.After(time.Duration(s.batchDelaySeconds) * time.Second):
```
Comment (bogdan-rosianu) [inline storage/leveldb/leveldb.go:92]: use ticker instead of time after?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: storage/leveldb/leveldb.go (hunk 3 above)
Comment (iulianpascalau) [reply storage/leveldb/leveldb.go:84]: right, removed logDebug pointer in order to keep only one log pointer on the package

File: storage/leveldb/leveldbSerial.go (hunk 4 above)
Comment (iulianpascalau) [reply storage/leveldb/leveldbSerial.go:84]: removed logdebug, kept only log pointer

File: storage/leveldb/leveldb.go (hunk 5 above)
Comment (iulianpascalau) [reply storage/leveldb/leveldb.go:92]: done in both implementations :1st_place_medal:

File: storage/leveldb/leveldbSerial_test.go [hunk 6]
```diff
@@ -47,21 +49,49 @@ func TestSerialDB_GetErrorOnFail(t *testing.T) {
 	assert.NotNil(t, err)
 }
 
-func TestSerialDB_CallsNotBlockingAfterCloseOrDestroy(t *testing.T) {
+func TestSerialDB_MethodCallsAfterCloseOrDestroy(t *testing.T) {
+	t.Parallel()
+
+	t.Run("when closing", func(t *testing.T) {
```
Comment (AdoAdoAdo) [inline storage/leveldb/leveldbSerial_test.go:55]: add maybe run in parallel as well

**Review by sasurobert — APPROVED** [review]

File: storage/leveldb/leveldbSerial_test.go (hunk 6 above)
Comment (iulianpascalau) [reply storage/leveldb/leveldbSerial_test.go:55]: done

File: state/peerAccountsDB.go [hunk 7]
```diff
@@ -90,11 +114,7 @@ func (adb *PeerAccountsDB) SnapshotState(rootHash []byte) {
 
 	go func() {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/peerAccountsDB.go:115]: reverted :1st_place_medal:

**Review by bogdan-rosianu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**